### PR TITLE
Add compatibility with Mesa-driven GPUs

### PR DESCRIPTION
**Pull request by 10110111**
This set of changes makes the demo work for on my two different Intel HD Graphics GPUs on Linux driven by Mesa `i965` driver, in addition to nVidia GPU driven by binary `nvidia` driver, which worked before.
To achieve this, the demo is switched to OpenGL Core Profile, framebuffer formats are chosen those which are guaranteed to exist by the OpenGL 3.3 specification, and an unnecessary GLSL extension is unused.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -86,19 +87,22 @@ output/Debug/atmosphere_test: \
 	$(GPP) $^ -o $@
 
 output/Release/atmosphere_integration_test: \
+    output/Debug/glad/src/glad.o \
```
Comment (ebruneton) [inline Makefile]: This should be output/Release/

File: Makefile [hunk 2]
```diff
@@ -29,11 +29,11 @@
 GPP := g++
 GPP_FLAGS := -Wall -Wmain -pedantic -pedantic-errors -std=c++11
 INCLUDE_FLAGS := \
-    -I. -Iexternal -Iexternal/dimensional_types -Iexternal/progress_bar
+    -I. -Iglad/include -Iexternal -Iexternal/dimensional_types -Iexternal/progress_bar
```
Comment (ebruneton) [inline Makefile]: Can you move the glad files to external using
  external/glad/include/glad.h
  extenal/glad/src/glad.cc
  extenal/glad/regenerate.sh

File: Makefile [hunk 3]
```diff
@@ -49,6 +49,7 @@ all: lint doc test integration_test demo
 # <regex>.
 lint: $(HEADERS) $(SOURCES)
 	cpplint --exclude=tools/docgen_main.cc \
+            --exclude=glad/src/glad.cc \
```
Comment (ebruneton) [inline Makefile]: This can be removed after you move the glad files.

File: atmosphere/demo/demo.cc [hunk 4]
```diff
@@ -136,7 +140,26 @@ Demo::Demo(int viewport_width, int viewport_height) :
...
+  glBindVertexArray(vao_);
+  glGenBuffers(1, &vbo_);
+  glBindBuffer(GL_ARRAY_BUFFER, vbo_);
+  const GLfloat vertices[]= {
+      -1.0, -1.0, 0.0, 1.0,
+      +1.0, -1.0, 0.0, 1.0,
+      -1.0, +1.0, 0.0, 1.0,
+      +1.0, +1.0, 0.0, 1.0,
+  };
+  constexpr int coordsPerVertex = 4;
```
Comment (ebruneton) [inline atmosphere/demo/demo.cc]: Please use kCoordsPerVertex (https://google.github.io/styleguide/cppguide.html#Constant_Names).

File: atmosphere/demo/demo.cc [hunk 5]
```diff
@@ -136,7 +140,26 @@ Demo::Demo(int viewport_width, int viewport_height) :
     INSTANCES[glutGetWindow()]->HandleMouseWheelEvent(dir);
   });
 
+  glGenVertexArrays(1, &vao_);
+  glBindVertexArray(vao_);
+  glGenBuffers(1, &vbo_);
+  glBindBuffer(GL_ARRAY_BUFFER, vbo_);
+  const GLfloat vertices[]= {
```
Comment (ebruneton) [inline atmosphere/demo/demo.cc]: Missing space between [] and =

File: atmosphere/demo/demo.cc [hunk 6]
```diff
@@ -136,7 +140,26 @@ Demo::Demo(int viewport_width, int viewport_height) :
...
+      -1.0, -1.0, 0.0, 1.0,
+      +1.0, -1.0, 0.0, 1.0,
+      -1.0, +1.0, 0.0, 1.0,
+      +1.0, +1.0, 0.0, 1.0,
+  };
+  constexpr int coordsPerVertex = 4;
+  glBufferData(GL_ARRAY_BUFFER, sizeof vertices, vertices, GL_STATIC_DRAW);
+  constexpr GLuint attribIndex = 0;
```
Comment (ebruneton) [inline atmosphere/demo/demo.cc]: kAttribIndex

File: atmosphere/demo/demo.h [hunk 7]
```diff
@@ -87,6 +90,8 @@ class Demo {
   bool show_help_;
 
   std::unique_ptr<Model> model_;
+  GLuint vao_;
+  GLuint vbo_;
```
Comment (ebruneton) [inline atmosphere/demo/demo.h]: Please use more explicit names, for instance fullscreen_quad_vao_ (https://google.github.io/styleguide/cppguide.html#General_Naming_Rules).

File: atmosphere/model.cc [hunk 8]
```diff
@@ -448,11 +447,9 @@ GLuint NewTexture3d(int width, int height, int depth, GLenum format,
   glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
   glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_WRAP_R, GL_CLAMP_TO_EDGE);
   glBindBuffer(GL_PIXEL_UNPACK_BUFFER, 0);
-  GLenum internal_format = format == GL_RGBA ?
-      (half_precision ? GL_RGBA16F : GL_RGBA32F) :
-      (half_precision ? GL_RGB16F : GL_RGB32F);
+  GLenum internal_format = half_precision ? GL_RGBA16F : GL_RGBA32F;
   glTexImage3D(GL_TEXTURE_3D, 0, internal_format, width, height, depth, 0,
-      format, GL_FLOAT, NULL);
+      GL_RGBA, GL_FLOAT, NULL);
```
Comment (ebruneton) [inline atmosphere/model.cc]: I think forcing RGBA is a big drawback here, it increases the GPU memory needs by 33% when this is not always necessary. Could you use 'format' by default and fallback to RGBA only if you get a framebuffer completness error?

File: atmosphere/model.h [hunk 9]
```diff
@@ -322,6 +323,8 @@ class Model {
       bool blend,
       unsigned int num_scattering_orders);
 
+  GLuint vao_;
```
Comment (ebruneton) [inline atmosphere/model.h]: Same comment as above, please use more explicit names.

File: atmosphere/reference/model_test.cc [hunk 10]
```diff
@@ -566,12 +571,32 @@ with the GPU program, and then read back the framebuffer pixels.
...
 
     glViewport(0, 0, kWidth, kHeight);
-    glBegin(GL_TRIANGLE_STRIP);
-    glVertex4f(-1.0, -1.0, 0.0, 1.0);
-    glVertex4f(+1.0, -1.0, 0.0, 1.0);
-    glVertex4f(-1.0, +1.0, 0.0, 1.0);
-    glVertex4f(+1.0, +1.0, 0.0, 1.0);
-    glEnd();
+    {
+        GLuint vao;
```
Comment (ebruneton) [inline atmosphere/reference/model_test.cc]: Idem.

File: atmosphere/reference/model_test.cc [hunk 11]
```diff
@@ -566,12 +571,32 @@ with the GPU program, and then read back the framebuffer pixels.
...
+        GLuint vbo;
+        glGenBuffers(1, &vbo);
+        glBindBuffer(GL_ARRAY_BUFFER, vbo);
+        const GLfloat vertices[] = {
+            -1.0, -1.0, 0.0, 1.0,
+            +1.0, -1.0, 0.0, 1.0,
+            -1.0, +1.0, 0.0, 1.0,
+            +1.0, +1.0, 0.0, 1.0,
+        };
+        constexpr int coordsPerVertex = 4;
```
Comment (ebruneton) [inline atmosphere/reference/model_test.cc]: kCoordsPerVertex

File: atmosphere/reference/model_test.cc [hunk 12]
```diff
@@ -566,12 +571,32 @@ with the GPU program, and then read back the framebuffer pixels.
...
+        glBufferData(GL_ARRAY_BUFFER, sizeof vertices, vertices,
+                     GL_STATIC_DRAW);
+        constexpr GLuint attribIndex = 0;
```
Comment (ebruneton) [inline atmosphere/reference/model_test.cc]: kAttribIndex

File: text/TextRenderer.cc [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "text/TextRenderer.h"
```
Comment (ebruneton) [inline text/TextRenderer.cc]: Can you rename TextRenderer.h/.cc to text_renderer.h/.cc for consistency with the other files?

File: text/TextRenderer.cc [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+#include "font.inc"
+
+void TextRenderer::setupTexture() {
+    glGenTextures(1, &fontTexture);
```
Comment (ebruneton) [inline text/TextRenderer.cc]: Please use onyl 2 spaces for indentation (for consistency with the other files).

File: text/TextRenderer.cc [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    constexpr GLuint attribIndex = 0;
+    constexpr int coordsPerVertex = 2;
+    glVertexAttribPointer(attribIndex, coordsPerVertex, GL_FLOAT, false, 0, 0);
+    glEnableVertexAttribArray(attribIndex);
+    glBindVertexArray(0);
+}
+
+void TextRenderer::setupProgram() {
+    const auto vertexShader = glCreateShader(GL_VERTEX_SHADER);
+    const char*const vertexShaderSrc = R"(#version 330
```
Comment (ebruneton) [inline text/TextRenderer.cc]: kVertexShaderSrc

File: text/TextRenderer.cc [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+void main()
+{
+    gl_Position = mvp*vertex;
+    texCoord = texCoordIn+charSizeInTexture*(vec2(vertex.x, 1-vertex.y));
+}
+)";
+    glShaderSource(vertexShader, 1, &vertexShaderSrc, nullptr);
+    glCompileShader(vertexShader);
+    const auto fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+    const char*const fragmentShaderSrc = R"(#version 330
```
Comment (ebruneton) [inline text/TextRenderer.cc]: kFragmentShaderSrc

File: text/TextRenderer.cc [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+    glLinkProgram(program);
+    glDetachShader(program, fragmentShader);
+    glDeleteShader(fragmentShader);
+    glDetachShader(program, vertexShader);
+    glDeleteShader(vertexShader);
+}
+
+void TextRenderer::drawChar(char c, int x, int y,
+                            int viewportWidth, int viewportHeight) {
+    if (c<0x20 || c>0x7e) c = '?';
```
Comment (ebruneton) [inline text/TextRenderer.cc]: Please use spaces around operators, throughout this file ("c < 0x20", "((c >> 4) - 2) / charH", etc)

File: text/TextRenderer.h [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef TEXT_TEXTRENDERER_H_
+#define TEXT_TEXTRENDERER_H_
+
+#include <glad/glad.h>
+
+class TextRenderer {
+    GLuint fontTexture;
```
Comment (ebruneton) [inline text/TextRenderer.h]: Can you make these fields private, use a trailing underscore (https://google.github.io/styleguide/cppguide.html#Variable_Names), and move them after the public members section (https://google.github.io/styleguide/cppguide.html#Declaration_Order)

File: text/TextRenderer.h [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    GLuint vao, vbo;
+    GLuint program;
+    GLfloat color[3];
+
+    void setupTexture();
```
Comment (ebruneton) [inline text/TextRenderer.h]: Can you use uppercase for the first letter of these method names (https://google.github.io/styleguide/cppguide.html#Function_Names)?

File: Makefile (hunk 3 above)
Comment (10110111) [reply Makefile]: This doesn't seem to work. To compile `glad.cc` I have to include `external/glad/src` to `DIRS`. But then `cpplint` tries to check `glad.cc`, which is undesired.

File: text/TextRenderer.h (hunk 18 above)
Comment (10110111) [reply text/TextRenderer.h]: Well they are already private, but OK, I'll move and rename them.

File: text/TextRenderer.cc (hunk 17 above)
Comment (10110111) [reply text/TextRenderer.cc]: Any way to double-check this? CppLint (on Kubuntu 14.04) seems to be quite happy with current state...

**Comment by 10110111** [comment]
I've addressed your concerns. Would you like me to squash these fixes into the original commits to avoid littering git-log output with style-fix noise etc.?

**Comment by 10110111** [comment]
Ah, yeah, I didn't touch the Windows-related files because I can't test them (don't have Windows with its build system handy).

**Comment by ebruneton** [comment]
Thanks for your changes, I will squash and merge them. I can't test on Windows either, unfortunately.

**Pull request merged**